File: packages/cucumber-universal-tests/features/step_definitions/circumference_steps.rb
require "json"

FIXTURE_PATH = File.expand_path(
  "../../fixtures/circumference.structure.json",
  __dir__
)

Given("the historical Sigillum data set is loaded") do
  @sigillum_dataset = JSON.parse(File.read(FIXTURE_PATH))
end

When("I inspect the circumference") do
  @circumference = @sigillum_dataset.fetch("circumference")
end

Then("it contains {int} chambers") do |expected|
  actual = @circumference.fetch("chambers").length
  raise "expected #{expected} chambers, got #{actual}" unless actual == expected
end

Then("each chamber spans {int} degrees") do |expected|
  @circumference.fetch("chambers").each do |chamber|
    actual = chamber.fetch("endDegree") - chamber.fetch("startDegree")
    raise "expected chamber span #{expected}, got #{actual}" unless actual == expected
  end
end

Then("the circumference spans {int} degrees") do |expected|
  actual = @circumference.fetch("chambers").sum do |chamber|
    chamber.fetch("endDegree") - chamber.fetch("startDegree")
  end
  raise "expected circumference span #{expected}, got #{actual}" unless actual == expected
end

When("I group the circumference into companies") do
  @companies = @sigillum_dataset
    .fetch("circumference")
    .fetch("chambers")
    .group_by { |chamber| chamber.fetch("company") }
end

Then("there are {int} companies") do |expected|
  actual = @companies.length
  raise "expected #{expected} companies, got #{actual}" unless actual == expected
end

Then("each company contains {int} chambers") do |expected|
  invalid = @companies.reject { |_company, chambers| chambers.length == expected }
  raise "company chamber counts did not equal #{expected}: #{invalid.keys}" unless invalid.empty?
end

File: packages/cucumber-universal-tests/features/steps/circumference_steps.py
import json
from pathlib import Path

from behave import given, then, when


FIXTURE_PATH = (
    Path(__file__).resolve().parents[2]
    / "fixtures"
    / "circumference.structure.json"
)


@given("the historical Sigillum data set is loaded")
def load_historical_sigillum(context):
    with FIXTURE_PATH.open("r", encoding="utf-8") as handle:
        context.sigillum_dataset = json.load(handle)


@when("I inspect the circumference")
def inspect_circumference(context):
    context.circumference = context.sigillum_dataset["circumference"]


@then("it contains {count:d} chambers")
def assert_chamber_count(context, count):
    assert len(context.circumference["chambers"]) == count


@then("each chamber spans {degrees:d} degrees")
def assert_chamber_span(context, degrees):
    for chamber in context.circumference["chambers"]:
        span = chamber["endDegree"] - chamber["startDegree"]
        assert span == degrees


@then("the circumference spans {degrees:d} degrees")
def assert_total_span(context, degrees):
    total = sum(
        chamber["endDegree"] - chamber["startDegree"]
        for chamber in context.circumference["chambers"]
    )
    assert total == degrees


@when("I group the circumference into companies")
def group_circumference(context):
    groups = {}
    for chamber in context.sigillum_dataset["circumference"]["chambers"]:
        groups.setdefault(chamber["company"], []).append(chamber)
    context.companies = groups


@then("there are {count:d} companies")
def assert_company_count(context, count):
    assert len(context.companies) == count


@then("each company contains {count:d} chambers")
def assert_company_size(context, count):
    assert all(len(chambers) == count for chambers in context.companies.values())
